### DISCUSSION TITLE
Will this emulate on Mobile Devices?

### DISCUSSION THREAD

**Discussion by WindowsTV** [ruffle-rs/ruffle#862] [General]
Hi, simple question, can Ruffle emulate SWF files on Mobile Devices such as iOS and Android?

**Comment by ROBERT-MCDOWELL** [comment]
as long as the browser will support WebAssembly I guess yes.

**Comment by Dinnerbone** [comment]
Per [the wiki](https://github.com/ruffle-rs/ruffle/wiki), the following mobile browsers have been tested and confirmed to work:

> Safari on iOS
> Chrome on Android
> Firefox on Android

Unfortunately, it can be difficult to play games that do not know about touch and rely on keyboard/mouse. We current treat touch as if it were a mouse, and this is enough for simple animations, menus and games, but a possible future goal could be to provide an on-screen virtual gamepad that can mimic keyboard input.

**Reply by WindowsTV** [reply]
Oh, interesting! The reason I asked is because the demo's page wont load on my iPhone. I wonder if I'm doing something wrong.

Thanks dinnerbone!

**Reply by desertking** [reply]
> Per [the wiki](https://github.com/ruffle-rs/ruffle/wiki), the following mobile browsers have been tested and confirmed to work:
> 
> > Safari on iOS
> > Chrome on Android
> > Firefox on Android
> 
> Unfortunately, it can be difficult to play games that do not know about touch and rely on keyboard/mouse. We current treat touch as if it were a mouse, and this is enough for simple animations, menus and games, but a possible future goal could be to provide an on-screen virtual gamepad that can mimic keyboard input.

This is a nice idea. But I see it far in the future. In the best case the player can detect the keys used itself and only insert those instead of a big keyboard or you have an api so the website owner can manually visualize them "W,A,S,D" for example to run around like on the keyboard and a "joystick" to aim and shoot if we talk about an shooter which uses that for moving and the mouse for aiming.

**Reply by Poopooracoocoo** [reply]
@WindowsTV 
> Oh, interesting! The reason I asked is because the demo's page wont load on my iPhone. I wonder if I'm doing something wrong.

Are you still unable to load the Ruffle demo page on your iPhone? It works for me on Safari for iPhone (running iOS 14 (and it did when it ran iOS 13)
(ah the golden puffle from cp... the memories)

**Comment by ROBERT-MCDOWELL** [comment]
puffin browser is already a good example of how they manage the virtual mouse and joystick

**Reply by WindowsTV** [reply]
Puffin indeed is good. When I learned about this emulator I was thinking about sourcing it out to run some flash games I have on mobile...I was having issues on the demo page hence my initial question.

**Reply by desertking** [reply]
Especially many flashgames are like 800x600 and scale to the middle you have enough space in landscape mode to the left and the right to enter a joystick for moving and some keys to the right ;)

*when in fullscreen, of course

**Comment by Herschel** [comment]
The demo page should work on mobile; if it isn't, could you file an issue with the specific device and browser you are using so that we can look into it? Thank you!
